File: backend/src/svap/stages/stage4c_assessment.py
"""
Stage 4C: Quality Assessment from Structural Findings

For each policy with completed structural research, evaluates every taxonomy
quality against the accumulated structural findings. Each assessment cites
specific findings as evidence.

Also syncs results to the legacy policy_scores table for backward compatibility
with stages 5 and 6.

Input:  structural_findings, taxonomy, research_sessions
Output: quality_assessments, policy_scores (backward compat)
"""

import hashlib

from svap.bedrock_client import BedrockClient
from svap.storage import SVAPStorage

ASSESSMENT_SYSTEM = (
    "You are assessing whether a structural vulnerability quality is present in a "
    "policy based on specific, cited structural findings. Be conservative — a quality "
    "is present only if findings directly support it. Cite specific finding IDs."
)


def run(storage: SVAPStorage, client: BedrockClient, run_id: str, config: dict):
    """Execute quality assessment against accumulated structural findings."""
    print("Stage 4C: Quality Assessment from Findings")
    storage.log_stage_start(run_id, 42)  # 42 = stage 4c

    try:
        taxonomy = storage.get_approved_taxonomy()
        if not taxonomy:
            print("  No taxonomy found. Run Stage 2 first.")
            storage.log_stage_complete(run_id, 42, {"policies_assessed": 0})
            return

        # Get policies with completed research
        sessions = storage.get_research_sessions(run_id, status="findings_complete")
        if not sessions:
            print("  No policies with completed research. Run Stage 4B first.")
            storage.log_stage_complete(run_id, 42, {"policies_assessed": 0})
            return

        assessed = 0
        for session in sessions:
            policy_id = session["policy_id"]
            findings = storage.get_structural_findings(run_id, policy_id)

            if not findings:
                print(f"  No findings for policy {policy_id}, skipping")
                continue

            policies = storage.get_policies()
            policy = next((p for p in policies if p["policy_id"] == policy_id), None)
            policy_name = policy["name"] if policy else policy_id

            print(f"  Assessing: {policy_name} ({len(findings)} findings)")

            findings_text = _format_findings(findings)

            for quality in taxonomy:
                assessment = _assess_quality(
                    client, quality, findings_text, findings, policy_id, policy_name, run_id
                )
                storage.upsert_quality_assessment(run_id, assessment)

                # Backward compat: sync to policy_scores
                present_bool = assessment["present"] == "yes"
                storage.insert_policy_score(
                    run_id, policy_id, quality["quality_id"],
                    present=present_bool,
                    evidence=assessment.get("rationale", ""),
                )

            storage.update_research_session(session["session_id"], "assessment_complete")
            storage.update_policy_lifecycle(policy_id, "fully_assessed")
            assessed += 1
            print(f"    Assessed {len(taxonomy)} qualities")

        storage.log_stage_complete(run_id, 42, {
            "policies_assessed": assessed,
            "qualities_per_policy": len(taxonomy),
        })
        print(f"  Assessment complete: {assessed} policies.")

    except Exception as e:
        storage.log_stage_failed(run_id, 42, str(e))
        raise


def _format_findings(findings: list[dict]) -> str:
    """Format findings into a readable text block for the LLM."""
    parts = []
    for f in findings:
        dim_name = f.get("dimension_name") or f.get("dimension_id") or "Unknown"
        parts.append(
            f"[{f['finding_id']}] ({dim_name}, {f.get('confidence', 'medium')} confidence)\n"
            f"  {f['observation']}\n"
            f"  Source: {f.get('source_citation', 'N/A')}"
        )
    return "\n\n".join(parts)


def _assess_quality(
    client: BedrockClient,
    quality: dict,
    findings_text: str,
    findings: list[dict],
    policy_id: str,
    policy_name: str,
    run_id: str,
) -> dict:
    """Assess one quality against the accumulated findings."""
    prompt = client.render_prompt(
        "stage4c_assess_quality.txt",
        quality_id=quality["quality_id"],
        quality_name=quality["name"],
        quality_definition=quality["definition"],
        quality_recognition_test=quality["recognition_test"],
        policy_name=policy_name,
        findings_text=findings_text,
    )

    result = client.invoke_json(prompt, system=ASSESSMENT_SYSTEM, temperature=0.1, max_tokens=1000)

    assessment_id = hashlib.sha256(
        f"{run_id}:{policy_id}:{quality['quality_id']}".encode()
    ).hexdigest()[:12]

    # Validate finding_ids against actual findings
    cited_ids = result.get("finding_ids", [])
    valid_ids = {f["finding_id"] for f in findings}
    validated_ids = [fid for fid in cited_ids if fid in valid_ids]

    return {
        "assessment_id": assessment_id,
        "policy_id": policy_id,
        "quality_id": quality["quality_id"],
        "taxonomy_version": run_id,
        "present": result.get("present", "uncertain"),
        "evidence_finding_ids": validated_ids,
        "confidence": result.get("confidence", "medium"),
        "rationale": result.get("reasoning", ""),
    }

File: backend/shared/src/stages/stage4c_assessment.rs
//! Stage 4C: Quality Assessment from Structural Findings

use serde_json::json;
use sha2::{Digest, Sha256};
use tokio_postgres::Client;
use tracing::info;

use crate::bedrock::BedrockClient;
use crate::db;
use crate::types::{
    Config, Policy, QualityAssessment, ResearchSession, StructuralFinding, TaxonomyQuality,
};

const ASSESSMENT_SYSTEM: &str = "You are assessing whether a structural vulnerability quality is present in a policy based on specific, cited structural findings. Be conservative -- a quality is present only if findings directly support it.";

const STAGE4C_ASSESS_PROMPT: &str = include_str!("../../prompts/stage4c_assess_quality.txt");
type StageResult<T> = Result<T, Box<dyn std::error::Error + Send + Sync>>;
type SessionAssessment = (ResearchSession, Vec<StructuralFinding>, String);

struct AssessmentContext<'a> {
    db_client: &'a Client,
    bedrock: &'a BedrockClient,
    run_id: &'a str,
}

fn compute_hash(parts: &[&str]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(parts.join("|"));
    format!("{:x}", hasher.finalize())[..12].to_string()
}

fn taxonomy_fingerprint(taxonomy: &[TaxonomyQuality]) -> String {
    let mut ids: Vec<&str> = taxonomy.iter().map(|q| q.quality_id.as_str()).collect();
    ids.sort();
    let mut hasher = Sha256::new();
    hasher.update(ids.join(":"));
    format!("{:x}", hasher.finalize())[..12].to_string()
}

pub async fn run(
    db_client: &Client,
    bedrock: &BedrockClient,
    run_id: &str,
    config: &Config,
) -> StageResult<serde_json::Value> {
    info!("Stage 4C: Quality Assessment from Findings");
    db::log_stage_start(db_client, run_id, 42).await?;

    match run_inner(db_client, bedrock, run_id, config).await {
        Ok(result) => {
            db::log_stage_complete(db_client, run_id, 42, Some(&result)).await?;
            Ok(result)
        }
        Err(e) => {
            db::log_stage_failed(db_client, run_id, 42, &e.to_string()).await?;
            Err(e)
        }
    }
}

async fn run_inner(
    db_client: &Client,
    bedrock: &BedrockClient,
    run_id: &str,
    _config: &Config,
) -> StageResult<serde_json::Value> {
    let taxonomy = db::get_approved_taxonomy(db_client).await?;
    if taxonomy.is_empty() {
        return Ok(json!({"policies_assessed": 0}));
    }

    let sessions = assessable_sessions(db_client).await?;

    if sessions.is_empty() {
        return Ok(json!({"policies_assessed": 0}));
    }

    // Delta detection
    let tax_fp = taxonomy_fingerprint(&taxonomy);
    let stored_hashes = db::get_processing_hashes(db_client, 42).await?;
    let (sessions_to_assess, skipped) =
        changed_sessions(db_client, &sessions, &stored_hashes, &tax_fp).await?;

    if sessions_to_assess.is_empty() {
        return Ok(json!({"policies_assessed": 0, "skipped_unchanged": skipped}));
    }

    let all_policies = db::get_policies(db_client).await?;
    let context = AssessmentContext {
        db_client,
        bedrock,
        run_id,
    };
    let assessed = assess_sessions(&context, &taxonomy, &all_policies, &sessions_to_assess).await?;

    info!(
        "Assessment complete: {} policies ({} unchanged).",
        assessed, skipped
    );
    Ok(
        json!({"policies_assessed": assessed, "skipped": skipped, "qualities_per_policy": taxonomy.len()}),
    )
}

async fn assessable_sessions(db_client: &Client) -> StageResult<Vec<ResearchSession>> {
    let sessions_new = db::get_research_sessions(db_client, Some("findings_complete")).await?;
    let sessions_done = db::get_research_sessions(db_client, Some("assessment_complete")).await?;
    let mut seen = std::collections::HashSet::new();
    let mut sessions = Vec::new();
    for session in sessions_new.iter().chain(sessions_done.iter()) {
        if seen.insert(session.policy_id.clone()) {
            sessions.push(session.clone());
        }
    }
    Ok(sessions)
}

async fn changed_sessions(
    db_client: &Client,
    sessions: &[ResearchSession],
    stored_hashes: &std::collections::HashMap<String, String>,
    tax_fp: &str,
) -> StageResult<(Vec<SessionAssessment>, usize)> {
    let mut changed = Vec::new();
    let mut skipped = 0;

    for session in sessions {
        let findings = db::get_structural_findings(db_client, &session.policy_id).await?;
        if findings.is_empty() {
            continue;
        }
        let hash = assessment_hash(&findings, tax_fp);
        if stored_hashes.get(&session.policy_id).map(|s| s.as_str()) == Some(&hash) {
            skipped += 1;
        } else {
            changed.push((session.clone(), findings, hash));
        }
    }

    Ok((changed, skipped))
}

fn assessment_hash(findings: &[StructuralFinding], tax_fp: &str) -> String {
    let mut finding_ids: Vec<&str> = findings.iter().map(|f| f.finding_id.as_str()).collect();
    finding_ids.sort();
    compute_hash(&[&finding_ids.join(":"), tax_fp])
}

async fn assess_sessions(
    context: &AssessmentContext<'_>,
    taxonomy: &[TaxonomyQuality],
    all_policies: &[Policy],
    sessions: &[SessionAssessment],
) -> StageResult<usize> {
    let mut assessed = 0;
    for (session, findings, hash) in sessions {
        assess_session(context, taxonomy, all_policies, session, findings).await?;
        db::record_processing(
            context.db_client,
            42,
            &session.policy_id,
            hash,
            context.run_id,
        )
        .await?;
        assessed += 1;
    }
    Ok(assessed)
}

async fn assess_session(
    context: &AssessmentContext<'_>,
    taxonomy: &[TaxonomyQuality],
    all_policies: &[Policy],
    session: &ResearchSession,
    findings: &[StructuralFinding],
) -> StageResult<()> {
    let policy_name = policy_name(all_policies, session);
    info!("Assessing: {} ({} findings)", policy_name, findings.len());
    let findings_text = findings_text(findings);

    for quality in taxonomy {
        assess_quality(
            context,
            session,
            findings,
            quality,
            &policy_name,
            &findings_text,
        )
        .await?;
    }

    db::update_research_session(
        context.db_client,
        &session.session_id,
        "assessment_complete",
        None,
        None,
    )
    .await?;
    db::update_policy_lifecycle(context.db_client, &session.policy_id, "fully_assessed").await
}

fn policy_name(all_policies: &[Policy], session: &ResearchSession) -> String {
    all_policies
        .iter()
        .find(|p| p.policy_id == session.policy_id)
        .map(|p| p.name.clone())
        .unwrap_or_else(|| session.policy_id.clone())
}

fn findings_text(findings: &[StructuralFinding]) -> String {
    findings
        .iter()
        .map(|f| {
            let dim_name = f
                .dimension_name
                .as_deref()
                .unwrap_or(f.dimension_id.as_deref().unwrap_or("Unknown"));
            format!(
                "[{}] ({}, {} confidence)\n  {}\n  Source: {}",
                f.finding_id,
                dim_name,
                f.confidence,
                f.observation,
                f.source_citation.as_deref().unwrap_or("N/A")
            )
        })
        .collect::<Vec<_>>()
        .join("\n\n")
}

async fn assess_quality(
    context: &AssessmentContext<'_>,
    session: &ResearchSession,
    findings: &[StructuralFinding],
    quality: &TaxonomyQuality,
    policy_name: &str,
    findings_text: &str,
) -> StageResult<()> {
    let result = invoke_assessment(context.bedrock, quality, policy_name, findings_text).await?;
    let assessment = quality_assessment(context.run_id, session, findings, quality, &result)?;
    db::upsert_quality_assessment(context.db_client, context.run_id, &assessment).await?;
    db::insert_policy_score(
        context.db_client,
        context.run_id,
        &session.policy_id,
        &quality.quality_id,
        assessment.present == "yes",
        assessment.rationale.as_deref().unwrap_or(""),
    )
    .await
}

async fn invoke_assessment(
    bedrock: &BedrockClient,
    quality: &TaxonomyQuality,
    policy_name: &str,
    findings_text: &str,
) -> StageResult<serde_json::Value> {
    let prompt = BedrockClient::render_prompt(
        STAGE4C_ASSESS_PROMPT,
        &[
            ("quality_id", &quality.quality_id),
            ("quality_name", &quality.name),
            ("quality_definition", &quality.definition),
            ("quality_recognition_test", &quality.recognition_test),
            ("policy_name", policy_name),
            ("findings_text", findings_text),
        ],
    );
    bedrock
        .invoke_json(&prompt, ASSESSMENT_SYSTEM, Some(0.1), Some(1000))
        .await
}

fn quality_assessment(
    run_id: &str,
    session: &ResearchSession,
    findings: &[StructuralFinding],
    quality: &TaxonomyQuality,
    result: &serde_json::Value,
) -> StageResult<QualityAssessment> {
    let present = result
        .get("present")
        .and_then(|p| p.as_str())
        .unwrap_or("uncertain")
        .to_string();
    Ok(QualityAssessment {
        assessment_id: assessment_id(&session.policy_id, &quality.quality_id),
        run_id: run_id.to_string(),
        policy_id: session.policy_id.clone(),
        quality_id: quality.quality_id.clone(),
        taxonomy_version: Some(run_id.to_string()),
        present,
        evidence_finding_ids: Some(serde_json::to_string(&validated_finding_ids(
            result, findings,
        ))?),
        confidence: result
            .get("confidence")
            .and_then(|c| c.as_str())
            .unwrap_or("medium")
            .to_string(),
        rationale: result
            .get("reasoning")
            .and_then(|r| r.as_str())
            .map(String::from),
        created_at: String::new(),
    })
}

fn assessment_id(policy_id: &str, quality_id: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(format!("{}:{}", policy_id, quality_id));
    format!("{:x}", hasher.finalize())[..12].to_string()
}

fn validated_finding_ids(
    result: &serde_json::Value,
    findings: &[StructuralFinding],
) -> Vec<String> {
    let valid_ids: std::collections::HashSet<&str> =
        findings.iter().map(|f| f.finding_id.as_str()).collect();
    result
        .get("finding_ids")
        .and_then(|f| f.as_array())
        .into_iter()
        .flatten()
        .filter_map(|v| v.as_str())
        .filter(|id| valid_ids.contains(*id))
        .map(String::from)
        .collect()
}
